### PR TITLE
Proposal for timeout issue fix

### PR DESCRIPTION
**Pull request by zofy**
This is a potential fix for the issue #51.
 * BaseServer.connection_lost() notifies the reader that the
connection was lost.
 * TelnetWriter.close() was modified to avoid calling the close, when
it's already closed.

This is indeed just a quick proposal so that we can decide if this is a good direction. The main purpose of thi PR is to at least start a discussion about how to fix the timeout issue.

### CODE REVIEW

**Comment by jquast** [comment]
This looks ok, you tested it fixes your issue, then? I will accept this

**Comment by zofy** [comment]
Well, yes it fixes the issue with the read() function, but after a deeper look I wonder if this approach is the best. For example in other functions like: `readexactly()` or `readuntil()` timeout would cause raising of `asyncio.IncompleteReadError`, which kind of makes sense. So a question came to my mind whether we shouldn't do the same in case of `read()` as well as `readline()` functions.
Another approach would be to alter the `TelnetServer.on_timeout()` method to raise a `TimeoutError` and leave the solution of the situation to the user. WDYT?
To summarize the options:
 - return u'' after the timeout in all cases (`read()`, `readuntil()`, `readline()`, `readexactly()`)
 - raise `asyncio.IncompleteReadError` in case of timeout in all of the functions
 - raise custom `TimeoutError` in `TelnetServer.on_timeout()`
 
 Which option would you prefer?

**Comment by jquast** [comment]
I prefer the first, second if we have to, never the third, thank you for your consideration, I’m trying to make time to return to this project soon, best wishes

**Comment by zofy** [comment]
Please, consider this approach as well. Instead of returning empty string in the mentioned cases, we could as well cancel the task on the reader. What do you think about this approach? Maybe there are some caveats, but the advantage would be to not have to modify those `read*` functions. If you strongly disagree, I'll return to the previous solution.

**Comment by jquast** [comment]
This is a fine approach

**Comment by zofy** [comment]
Alright, thank you. This is my final PR. I guess, there's no reason to update the tests as there are already tests for the timeout.
Could you please review the PR?

**Comment by jquast** [comment]
this looks good, thank you, I will merge and release soon
